### ISSUE TITLE
For WTI submit run add governance/pacing to reduce flooding server with runs

### ISSUE THREAD

**Issue by lane55** [pc2ccs/pc2v9#764]
**Is your feature request related to a problem?** 

Yes, a type of DOS can be done when using WTI.
A small script can be run to submit hundreds of runs (as fast as pssible) 

**Feature Description**: 

Prevent a team from making too many submissions within a certain period of time. Implement a governor algorithm.

_The scope of this change is ONLY for WTI, so it should be enforced in the pc2 java API (specifically in the ServerConnection class)._
NOTE: the above comment, written in the original issue, is probably no longer accurate; the scope should include not only WTI but also any other mechanism for submitting (e.g. the Fat Team Client and command-line or programmatic submissions via the CLICS API).

_This should not be enforced in InternalController because when run as shadow the server should allow many runs as fast as possible, esp if Shadowing is restarted._
NOTE: the above comment, written in the original issue, may not be accurate/relevant; see the discussion below regarding dealing with Shadow submissions.

**Have you considered other ways to accomplish the same thing?** 
Human intervention is the only other way to deal with this type of problem (turn of the ability to submit runs when
the condition is first observed).

**Do you have any specific suggestions for how your feature would be ***implemented*** in PC^2?**

**Additional context**: 
This issue first came to light in the Spring 2023 PacNW online practice contest, which contained an interactive problem called "Secret Remainder" which asked for a program to repeatedly "guess a number" in the range 110..1,000,000.  With their advisor's permission, a team ran the following script:
```python
from time import sleep
import requests
import base64

for M in range(110, 1_000_000):

  code = f"print(int(input())%({M}))"
  bytes = code.encode('ascii')
  submission = base64.b64encode(bytes).decode('ascii')
  print(f"submitting {M}")

  result = [requests.post](http://requests.post/)(
    "http://contest.icpc-vcss.org/api/teams/run",
    json = {
      'mainFile': {
        'byteData': submission,
        'fileName':'soln.py'
      },
      'extraFiles':[],
      'testFile':None,
      'additionalTestFiles':[],
      'probName':"F: Secret Remainder",
      "language": "Python 3",
      "isTest": False,
      "osName": "idk"
    },
    headers = {
      'team_id':redacted
    }
  )
  sleep(1)
```
Basically, a program which makes about a million submissions, one second apart, by invoking the PC2 WTI "submit" endpoint.

**Comment by clevengr** [comment]
This appears to be a duplicate of Issue #702, with the only differences being minor details in comments (but not in the basic issue).  Either this issue or #702 should be marked as a duplicate and closed.  Note that relevant comments from whichever gets closed should be copied into the other issue.

**Comment by johnbrvc** [comment]
After thinking about the comment that Doug had made in this issue:

> This should not be enforced in InternalController because when
> run as shadow the server should allow many runs as fast as possible,
> esp if Shadowing is restarted

I think we can ignore that comment because I do not think it is necessary to worry about adding special code in our throttle algorithm(s) for shadows for a couple of reasons.
The "_time_" we will use on the submission would be the "_override_" time in the received submission, which is the _contest time_ (time since the contest started).  In the case of a replay, even though we may receive all submissions for an entire contest in less than a minute, each team's submissions will still have the _contest time_ spread out over the entire contest (be it 5 hours or 2 hours), so while we may get, say, 50 submissions from a team in a real-time minute, their contest time's (override times) will be spread out over the length of the original contest, so any throttle algorithm we implement should not be triggered.
If we are shadowing, this means we are not accepting submissions from team clients (that is, we are not accepting submissions from `pc2team`, **WTI** or the `CLICS submission EP`).  Therefore, we can set the tunable parameters for any throttling algorithm to  values that would never cause a submission to be rejected.  As @JoeTerlizzi  suggested, we can have a config setting, or settings, in the `pc2v9.ini` file (or something), that tells the throttling algorithm to either let all submissions through (don't throttle), or adjust the timing/submission count params so they are never triggered.

There's another case we did not consider last night and that is the case of "judge's" sample submissions, where all of the judge's submissions are typically made within a short time (a minute or two).  I don't think this is a problem either since typically, there may be half a dozen or so submissions for each problem.  We could again set the algorithm throttle params such that it permits 6 or (whatever the number is) of submissions per minute.  In addition, we could also not make the throttling algorithm apply if the submitter is not a team (we discussed this last night), or, if the submitter is not being shown on the scorebooard. 
 
All that being said, I don't think we have to worry about adding special code to determine if a submission is from the shadow or is a judge's submissions case - just implement the algorithm wherever in the code we decide to do that.

So, I think we don't even worry about the shadow case.

**Comment by clevengr** [comment]
> After thinking about the comment that Doug had made in this issue:
> 
> > This should not be enforced in InternalController because when
> > run as shadow the server should allow many runs as fast as possible,
> > esp if Shadowing is restarted
> 
> I think we can ignore that comment because I do not think it is necessary to worry about adding special code in our throttle algorithm(s) for shadows for a couple of reasons. The "_time_" we will use on the submission would be the "_override_" time in the received submission, which is the _contest time_ (time since the contest started).  In the case of a replay, even though we may receive all submissions for an entire contest in less than a minute, each team's submissions will still have the _contest time_ spread out over the entire contest (be it 5 hours or 2 hours), so while we may get, say, 50 submissions from a team in a real-time minute, their contest time's (override times) will be spread out over the length of the original contest, so any throttle algorithm we implement should not be triggered. 

But: we are (I think) talking about the desirability of invoking the throttling algorithm in **InternalController.submitJudgeRun(...)**, so that throttling is done on the _client_ side rather than the _server_ side (right?) But the _overrideTime_ doesn't get checked on the client side; it gets passed (in the **_RUN_SUBMISSION_** packet) to the server, which is where it gets checked/handled.  I agree that the throttling algorithm(s) themselves probably shouldn't be (shouldn't need to be) checking for "shadow/not shadow" -- but it appears to me that the **_submitJudgeRun()_** method in the InternalController _does_ need to do such a check -- it can't just blindly call the throttling algorithm on every submission, because that will cause shadow submissions to the server to get throttled.

In other words, I imagine that _submitJudgeRun()_ needs to be enhanced to be (something like) the following:
```
submitJudgeRun(...) {
  ...
  boolean accept = true;
  if (contest.getClientId().isTeam()) {
    ThrottleStrategy strategy = new ThrottleStrategy();  //implementation of the current throttling strategy/algorithm
    accept = strategy.accept(run);   //returns true if run should be accepted according to the current strategy, false if not
  }
  if (accept) {
    sendToLocalServer (new RunSubmissionPacket());
  } else {
    sendToLocalServer (new RunSubmissionRejectedPacket());  //new type of packet, indicating a submission was rejected
  }
```
That is, the throttling strategy implementation itself does _not_ need to worry about whether the run came from a shadow or a team -- but we _do_ need to check _somewhere_ for "who the run came from" -- so we can avoid throttling if it came from a shadow.  But, @johnbrvc : perhaps this is what you were saying and I misinterpreted it?

>If we are shadowing, this means we are not accepting submissions from team clients (that is, we are not accepting submissions from `pc2team`, **WTI** or the `CLICS submission EP`).  Therefore, we can set the tunable parameters for any throttling algorithm to  values that would never cause a submission to be rejected.  As [@JoeTerlizzi](https://github.com/JoeTerlizzi) suggested, we can have a config setting, or settings, in the `pc2v9.ini` file (or something), that tells the throttling algorithm to either let all submissions through (don't throttle), or adjust the timing/submission count params so they are never triggered.

I don't see that we need to do this, and I think it's confusing -- it mixes the notion of "execute an algorithm that decides whether throttling should be enforced for a given run" with the notion of "where that run came from".  I don't think we need to mix these two things.  I think a better approach is to define an _interface_ which throttling algorithms implement, and then provide different concrete _implementations_ of that interface which enforce their own rules (and do so independently of the original source of the run).   I think the decision about "whether to consider applying throttling" should be done in **_submitJudgeRun()_**  as described above, and then each throttling algorithm implementation contains code that decides whether to accept or reject (i.e., throttle) a run based on the past history of the submitter (which will always be a team because the **_submitJudgeRun()_** code has bypassed throttling for non-teams) and the specific algorithm implemented by the concrete implementation.

For example:
```
interface ThrottleStrategy {
  //this method returns whether or not to accept the specified run based on the concrete strategy class implementation
  boolean accept(run) ;
}
```
We could provide as many different concrete implementations of "ThrottleStrategy" as different ways we can think of to implement throttling.  For example:
```
/** This class implements throttling (rejects a submission) if the team has exceeded the "max per minute" limit. **
class MaxSubmissionsPerMinute implements ThrottleStrategy {
  IInternalContest contest ;
  int maxPerMinute ;
  public MaxPerMinute (InternalContest inContest, int maxAllowed) {
    this.contest = inContest ;
    this.maxPerMinute = maxAllowed ;
  }
  boolean accept (Run run) {
    int numSoFar = getNumberOfRunsSubmittedInTheLastMinute(contest, run.getClientId());
    if (numSoFar < maxPerMinute) 
      return true;
    else 
      return false ;
   }
```
We _could_  just "hard-code" a single implementation of _ThrottleStrategy_ into the system.  Alternatively,
the Administrator could be given the ability to select which concrete implementation should be used for a given contest -- for example, by having a YAML entry like
```
throttleStrategy : MaxSubmissionsPerMinute
```
and we could provide the ability to set the strategy parameters, like:
```
throttleStrategy.maxPerMinute : 10
```
We could also (SomeDay (tm)) display a list of the available ThrottleStrategy implementations on the Admin GUI and allow the Administrator to select the desired algorithm dynamically...

> 
> There's another case we did not consider last night and that is the case of "judge's" sample submissions, where all of the judge's submissions are typically made within a short time (a minute or two).  I don't think this is a problem either since typically, there may be half a dozen or so submissions for each problem.  We could again set the algorithm throttle params such that it permits 6 or (whatever the number is) of submissions per minute.  In addition, we could also not make the throttling algorithm apply if the submitter is not a team (we discussed this last night), 

The approach I've described (above) does exactly that -- but it separates the notion of "the submitter is a team" from the notion of the exact throttling algorithm.  To do otherwise means we have to require every implementation of throttling to take care to check for whether the run came from a team.  I don't think we want to do that.

> ...or, if the submitter is not being shown on the scorebooard.    

Hmmm... didn't think about this case... but it's trivial to add this test to add that to the test in **_submitJudgeRun()_**.

> All that being said, I don't think we have to worry about adding special code to determine if a submission is from the shadow or is a judge's submissions case - just implement the algorithm wherever in the code we decide to do that.

Again, I think we _do_ want to worry about where the submission came from -- but I think we _don't_ want to (or need to) do that _in the throttling algorithm_, as shown above.
